ENH: Move itk::PhasedArray3DSpecialCoordinatesImage to ITK core

dzenanz opened the pull request:
This requires ITK python package with itk::PhasedArray3DSpecialCoordinatesImage wrapping added in: https://github.com/InsightSoftwareConsortium/ITK/pull/4101.

Supersedes and closes #237.

dzenanz committed:
ENH: Move itk::PhasedArray3DSpecialCoordinatesImage to ITK core

Remove itk::PhasedArray3DSpecialCoordinatesImage wrapping from:
https://github.com/InsightSoftwareConsortium/ITK/pull/4101

Closes #234.

tbirdso commented:
Looks good to me.

Maybe @thewtex can comment on the ITK pre-release schedule?

dzenanz commented:
Converting to draft, so we don't accidentally merge it before packages are ready.

thewtex commented:
Excellent!

The wrapping for PhasedArray3DSpecialCoordinatesImage will be in itk-5.4rc2. Let's update this repository to v5.4rc1 first.

dzenanz committed:
ENH: Update CI and bump package version to 0.7.1

dzenanz committed:
COMP: Specify dependent remote module version as hashes instead of tags

dzenanz commented:
Converting to draft until I resolve Python packaging issues, which I can now reproduce locally.

dzenanz committed:
COMP: Use ComputeIndex instead of GetIndex to avoid a warning

The first error message was:

C:\Dev\ITKUltrasound\include\itkBackscatterImageFilter.hxx(107,33): warning C4996: 'itk::ImageConstIterator<TImage>::GetIndex': Please use `ComputeIndex()` instead, or use an iterator with index, like `ImageIteratorWithIndex`!

dzenanz committed:
ENH: Use latest versions of dependencies in CI

dzenanz committed:
BUG: Avoid undefined behavior

Indeed, the in the else branch transform was null.
Presumably, compiler optimized it away previously.
The offending statements are now removed, to be on the safe side.

The first warning was:

[133/210] Building CXX object test/CMakeFiles/UltrasoundTestDriver.dir/itkSliceSeriesSpecialCoordinatesImageTest.cxx.o
In file included from /home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/test/itkSliceSeriesSpecialCoordinatesImageTest.cxx:20:
/home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/include/itkSliceSeriesSpecialCoordinatesImage.h: In member function ‘void itk::SliceSeriesSpecialCoordinatesImage<TSliceImage, TTransform, TPixel, VDimension>::TransformContinuousIndexToPhysicalPoint(const itk::ContinuousIndex<TIndexRep, VDimension>&, itk::Point<TCoordRep, VDimension>&) const [with TCoordRep = double; TIndexRep = double; TSliceImage = itk::CurvilinearArraySpecialCoordinatesImage<unsigned char, 2>; TTransform = itk::Euler3DTransform<double>; TPixel = unsigned char; unsigned int VDimension = 3]’:
/home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/include/itkSliceSeriesSpecialCoordinatesImage.h:476:42: warning: ‘this’ pointer is null [-Wnonnull]
  476 |         point = transform->TransformPoint(point);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
In file included from /home/local/KHQ/dzenan.zukic/ITK/ITK-git/Modules/Core/Transform/include/itkMatrixOffsetTransformBase.h:598,
                 from /home/local/KHQ/dzenan.zukic/ITK/ITK-git/Modules/Core/Transform/include/itkRigid3DTransform.h:22,
                 from /home/local/KHQ/dzenan.zukic/ITK/ITK-git/Modules/Core/Transform/include/itkEuler3DTransform.h:22,
                 from /home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/include/itkUltrasoundImageFileReader.hxx:24,
                 from /home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/include/itkUltrasoundImageFileReader.h:81,
                 from /home/local/KHQ/dzenan.zukic/ITK/ITKUltrasound/test/itkSliceSeriesSpecialCoordinatesImageTest.cxx:22:
/home/local/KHQ/dzenan.zukic/ITK/ITK-git/Modules/Core/Transform/include/itkMatrixOffsetTransformBase.hxx:124:1: note: in a call to non-static member function ‘itk::MatrixOffsetTransformBase<TParametersValueType, VInputDimension, VOutputDimension>::OutputPointType itk::MatrixOffsetTransformBase<TParametersValueType, VInputDimension, VOutputDimension>::TransformPoint(const InputPointType&) const [with TParametersValueType = double; unsigned int VInputDimension = 3; unsigned int VOutputDimension = 3]’
  124 | MatrixOffsetTransformBase<TParametersValueType, VInputDimension, VOutputDimension>::TransformPoint(
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

dzenanz commented:
This is now rebased on top of #243.
